### ISSUE TITLE
fresh npm install stripe and error

### ISSUE THREAD

**Issue by rkoberg** [stripe/stripe-node#265]
Just importing stripe (```import StripeFactory from 'stripe';```) or using require shown in the README causes an error:
```
ERROR in ./~/stripe/lib/stripe.js
Module not found: Error: Cannot resolve module 'child_process' in /Users/rkoberg/Sites/fwi/node_modules/stripe/lib
 @ ./~/stripe/lib/stripe.js 24:11-35
```
I am using the fix mentioned here: https://github.com/webpack/webpack.js.org/issues/107

But can a maintainer fix this. The fix for your module is also at the link above, specifically:

> The proper course is the ask the module author to make the module browser compatible. The author can add browser: { fs: false, child_process: false } to the package.json to tell webpack that it's ok for the module to get an empty object for these modules.

**Comment by rkoberg** [comment]
Is this module strictly for a server side implementation? Or how do you set up for the browser?
Spoke too soon about a fix. Now I get the error (in the browser, not the server):
```
stripe.js?7e15:9
Uncaught TypeError: __webpack_require__(...).createServer is not a function
(anonymous function) @ stripe.js?7e15:9
(anonymous function) @ stripe.js?7e15:189
```

**Comment by floatingLomas** [comment]
Hi!  Yes, this module is only to be used server-side, as it requires using a secret key that should never be shared into the wild (on the web).

If you're looking to capture card details, [Stripe.js](https://stripe.com/docs/stripe.js) can help you do that.

**Comment by rkoberg** [comment]
Ah. I thought this was Stripe.js :) Is Stripe.js not an npm module? I can't seem to find it. I do see the docs say you should include it from the CDN. Is that the recommended way even if you are packaging things up with webpack?

**Comment by SlexAxton** [comment]
Yes. The only supported method of including stripe.js is via  https://js.stripe.com/v2/ (primarily for compliance reasons)

**Comment by rkoberg** [comment]
Thanks for the quick responses!

**Comment by SlexAxton** [comment]
We have stripe.js distributed across edge-servers around the world, so it should be very fast, but it also happens to be the case that stripe.js is rarely needed directly on page-load  (usually not until a `createToken` function needs to be invoked, or similar), so feel free to defer loading it until after other scripts if you're worried about performance.

**Comment by rkoberg** [comment]
Do you have any advice or examples for setting this up in webpack? I tried just including the script element (with and without defer="defer"), and next, in the webpack config:

```
    plugins: (() => {
      const plugins = [
...
        new webpack.ProvidePlugin({
          Stripe: "stripe",
          "window.Stripe": "stripe"
        }),
...
     ]
```
But window.Stripe is undefined. Any advice/examples on how to get this to work in a webpack environment?

**Comment by rkoberg** [comment]
Btw, I was getting a conflict (?) when trying to require('stripe') -- it brought in this module rather than the Stripe.js

**Comment by SlexAxton** [comment]
You can inject via just regular javascript, if that works.

```javascript
let cachedStripeJsObject = null;

const StripeJsProvider = (readyCallback) => {
  // return early if we've already done all the hard work
  if (cachedStripeJsObject) {
    return readyCallback(this, cachedStripeJsObject);
  }

  // Find a good place to inject it, before the first script seems like a logical choice since
  // we definitely know one exists since this script is running.
  const firstScript = document.getElementsByTagName('script')[0];

  // create a new script
  const script = document.createElement('script');

  // set it to stripe.js
  script.src = 'https://js.stripe.com/v2/';

  // might not be necessary depending on when you inject
  script.async = true;
  script.defer = true;

  // get a callback when it's done
  script.onload = () => {
    // delete the handler, just for memory reasons;
    script.onload = undefined;
    // Cache the object so we don't reinject
    cachedStripeJsObject = window.Stripe;
    // call the callback, which I've decided randomly gets the object sent via the global
    // that what you don't have to mention the global variable ever again.
    readyCallback.apply(this, window.Stripe);
  };

  // Error handling/timeouts can be handled here as well.

  // Inject the script into the dom, which sets off the loading/execution
  firstScript.parentElement.insertBefore(script, firstScript);
};

export default StripeJsProvider;
```

(this is all from memory, but it should be a pretty good start)

Then you can just pull in the stripe.js provider at earliest moment you think you can afford to load stripe.js, but then all subsequent usages don't require any network.

**Comment by rkoberg** [comment]
Hmm... will have to figure out a good way to deal with this. But, I have Stripe.js with:
```
  if (process.env.IS_BROWSER) {
    const $script = require('scriptjs'); // can't import this -- only works in browser
    $script.get('https://js.stripe.com/v2/', function () {
      console.log('Stripe', Stripe);// prints "function a(){}"
    });
  }
```
but not sure how this will work bundling, probably need webpack's require.ensure. Also, this is for an isomorphic app where it can render on the server and the client (browser). (no ```document``` on the server). Not sure if I would need it to render on the server. Perhaps a client requests a page from the server that needs something from Stripe.js? 

A little surprised there is not a ready-made solution for this.

**Comment by rkoberg** [comment]
I guess what I mean is for an isomorphic app I will need tests to check 'if browser' anywhere I need to use Stripe.js. I am wondering if I should just be dealing with API endpoints directly and forego Stripe.js. Curious how others have handled this issue.

**Comment by SlexAxton** [comment]
stripe.js is very specifically a browser library. It makes requests with CORS, and uses browser primitives in order to complete certain actions. You can switch between using stripe-node and stripe.js, but they very intentionally have different functionality. The browser is a lot more limited in what it can do since it's a publicly-visible runtime (you can't do anything that requires a private key).

I honestly can't think of a case where things make sense with isomorphic stripe.js - what would the server be doing? stripe.js does not do things onRender. It mostly just creates tokens after a user interacts with the page.

As far as I know, there's no way to use stripe.js on the serverside for tokenizations without sending credit card numbers directly to your server, which is the entire thing that stripe.js is trying to help you avoid. You need a much more stringent level of PCI certification in order to support that use-case.

Also, in order to be compliant with PCI DSS 3.1 stripe.js does a bit of work to make all requests back to the Stripe API from inside a hidden iframe in your page. Not to mention, doing any of this work directly against the API puts you in other scarier PCI compliance modes, which is why we require that you pull in stripe.js from our CDN. All of this is necessary, not-compatible-with-isomorphic-stuff-in-general, and probably all totally fine to not include on the server side.

> A little surprised there is not a ready-made solution for this.

> Curious how others have handled this issue.

I'm not unfamiliar with serverside/isomorphic javascript apps ("I did it before it was cool!"), but I honestly think it's just not an issue because the goal of stripe.js is fundamentally incompatible with running it on your server (the library exists so you don't run it on your server). We have people confused between the client-side and server-side libraries from time to time, but a lot of thought has gone into the current set of tools. I also don't think at the moment, a single isomorphic library for privileged serverside actions would be good to mix in with the client-side tokenization library for hitting the your serverside library. They perform two distinct tasks that would be more often confused if they were together.

**Comment by rkoberg** [comment]
OK. I was confused about the difference between Stripe.js and node-stripe. I thought it was the same library, and that the docs at https://stripe.com/docs/stripe.js were more fully realized at https://stripe.com/docs/api/node; that only certain calls would require the private key and others only needed the public key. For example, I thought I would be able to GET products and skus/variants on the client (rather than needing to proxy through my app server), but I see/understand now I can't. Apologies for my confusion. Fwiw, I think Stripe.js is a bad name for a browser/client library that can only parse a form to get a single use token. Stripe.js seems more like a name for a library/module that can do all things Stripe. Perhaps names like stripe-client.js and stripe-server.js would be more clear?

**Comment by aviadhahami** [comment]
Hey, I've got here doing the same mistake @rkoberg did by confusing `stripeJS` and `stripe-node` and I think there should be a clarification for that on the main `readme`. Especially for other stripe-wanderers that may get themselves into PCI-hell by using the node version accidentally.

One more thing, @SlexAxton , aren't we exposing a new kind of threat by calling `Stripe` globally (via `<script>`) ?
In this case a user with some chrome console knowledge can access the Stripe object with relative ease and perform API calls on our behalf (our = token holders).
I'm pretty sure you've thought this through, but this is just a thought.

**Comment by SlexAxton** [comment]
@aviadhahami that's one of the main reasons it's important not to put stripe-node in the browser. Stripe.js has very limited access with your publishable key. It can create tokens, but it can't charge them without sending it to your server, and passing along the secret key. So for the most part, people could annoy you by filling up your tokenization logs in the dashboard, but they couldn't charge a card or retrieve any information from the API on your behalf. Essentially, they can't do anything that they couldn't already do by just filling out your form with bogus information.

Just make sure no one knows your secret keys (live or test mode), and you should be good.

**Comment by SlexAxton** [comment]
I went ahead and updated the readme up top with a note about the difference between the libraries.

**Comment by floatingLomas** [comment]
> It can create tokens, but it can't charge them without sending it to your server, and passing along the secret key.

Just to add some clarity here for future readers: The secret key doesn't get 'passed along' from/by Stripe.js; it should never be put anywhere near a browser.  It should only ever be (stored safely on and) used by your server.  

Stripe.js can create tokens using your publishable key, but those tokens are absolutely worthless until they're 'combined' with your secret key on the server side to actually create the charges.

**Comment by SlexAxton** [comment]
Yes, sorry, that was an ambiguous sentence. I meant to imply that 'your server' passed along the secret key, not stripe.js. Thanks for helping clarify.

**Comment by acmitch** [comment]
@SlexAxton Can you give us an example on how to use/call the 'StripeJsProvider', for clarification purposes? It would be a big help.

**Comment by jlomas-stripe** [comment]
@acmitch It's really just a 'deferred loading callback', sort of similar to jQuery's [`.ready()`](https://api.jquery.com/ready/); you just need to add it into your async flow:

    Stripe; // == `undefined`; It hasn't loaded yet

    StripeJsProvider(function() {
        Stripe; // == the Stripe.js Object, because it's now loaded.
    });

Does that make sense?

**Comment by acmitch** [comment]
@jlomas-stripe So something like this? Sorry, not proficient in Javascript.

```javascript
import React from "react";
import StripeJsProvider from "../../../providers/stripeProvider";
let Stripe; // == `undefined`; It hasn't loaded yet

class ShowProject extends React.Component {
    constructor(props, context) {
        StripeJsProvider(() => {
            Stripe; // == the Stripe.js Object, because it's now loaded.
        });
       Stripe.setPublishableKey('pk_test_erWk96SmDKGayGBW4QNP59qm');
    }
}
```

**Comment by jlomas-stripe** [comment]
@acmitch Ahhh, no, not quite like that.  I'm pretty sure it's going to have to be done differently if you're using it in React components.  This might be useful: https://davidwalsh.name/step-step-guide-stripe-payments-react

**Comment by acmitch** [comment]
@janelle-stripe Awesome! The example in the link will do nicely. Appreciate the help.

**Comment by jlomas-stripe** [comment]
No problemo! 👍
